tests/unit/s2n_build_test.c

```diff
@@ -0,0 +1,228 @@
+    } else {
+        return -1;
+    }
+    if (li->minor_version != -1) {
+        sprintf(version_string_out, ".%d", li->minor_version);
+        version_string_out = &version_string_out[2];
+    }
+    if (li->patch_version != -1) {
+        sprintf(version_string_out, ".%d", li->patch_version);
+    }
+    return 0;
+}
+
+int extract_libcrypto_info_from_s2n_libcrypto(char *s2n_libcrypto_copy, struct libcrypto_info *li)
+{
+    if (li == NULL)
+        return -1;
+    if (strlen(s2n_libcrypto_copy) == 0)
+        return -1;
+    li->name = strtok(s2n_libcrypto_copy, "-");
+    char *rest = strtok(NULL, "");
+
+    if (rest != NULL && isdigit(rest[0])) {
+        char *saved = NULL;
+        li->major_version = number_from_version(strtok(rest, ".-"));
+        if (li->major_version != -1) {
+            saved = strtok(NULL, ".-");
+            li->minor_version = number_from_version(saved);
+        }
+        if (li->minor_version != -1) {
+            saved = strtok(NULL, ".-");
+            li->patch_version = number_from_version(saved);
+        }
+        if (li->patch_version != -1) {
+            rest = strtok(NULL, "");
+        } else {
+            rest = saved;
+        }
+    }
+
+    if (rest != NULL) {
+        li->is_fips = (strstr(rest, "fips") != NULL);
+    }
+    return 0;
+}
+
+int extract_libcrypto_info_from_s2n_build_preset(char *s2n_build_preset_copy, struct libcrypto_info *li)
+{
+    if (li == NULL)
+        return -1;
+    if (strlen(s2n_build_preset_copy) == 0)
+        return -1;
+    li->name = strtok(s2n_build_preset_copy, "-_");
+    char *rest = strtok(NULL, "");
+    if (rest != NULL && isdigit(rest[0])) {
+        li->major_version = number_from_version(strtok(rest, ".-"));
+        if (li->major_version != -1) {
+            li->minor_version = number_from_version(strtok(NULL, ".-"));
+        }
+        if (li->minor_version != -1) {
+            li->patch_version = number_from_version(strtok(NULL, "-_"));
+        }
+        rest = strtok(NULL, "");
+    }
+
+    if (rest != NULL) {
+        li->is_fips = (strstr(rest, "fips") != NULL);
+    }
+    return 0;
+}
+
+int main()
+{
+    /* There are three different specifications of libcrypto which must be in sync.
+     *  1 The #include headers (-I) (checked with macros, e.g. OPENSSL_VERSION_TEXT and OPENSSL_IS_AWSLC)
+     *  2 The (dynamically) linked library (-L) (checked by calling function e.g. SSLeay_version)
+     *  3 The intent of the CI (declared by environment variables S2N_BUILD_PRESET and S2N_LIBCYPTO)
+     * BEGIN_TEST calls s2n_init which checks that (1) and (2) match. */
+    BEGIN_TEST();
+    /* The rest of the test checks that (3) matches (1) or (2). */
+
+    /* If both S2N_BUILD_PRESET and S2N_LIBCRYPTO are not defined, we aren't in a CI, end the test. */
+    const char *s2n_build_preset = getenv("S2N_BUILD_PRESET");
+    const char *s2n_libcrypto = getenv("S2N_LIBCRYPTO");
+    if (s2n_build_preset == NULL && s2n_libcrypto == NULL) {
```

Comment:
Unfortunately, there are some places in the CI where we only define S2N_LIBCRYPTO and not S2N_BUILD_PRESET. We could just rely on S2N_LIBCRYPTO (since it is set based on S2N_BUILD_PRESET, see: codebuild/bin/s2n_set_build_preset.sh). Except that's part of the test here (making sure it was set properly).

Reply:
Given that as a constraint, there might be a better way to arrange the logic here.